numpy/typing/_callable.py

```diff
@@ -0,0 +1,140 @@
+"""
+A module with various ``typing.Protocol`` subclasses that implement
+the ``__call__`` magic method.
+
+See the `Mypy documentation`_ on protocols for more details.
+
+.. _`Mypy documentation`: https://mypy.readthedocs.io/en/stable/protocols.html#callback-protocols
+
+"""
+
+import sys
+from typing import Union, TypeVar, overload, Any
+
+from numpy import (
+    _BoolLike,
+    _FloatLike,
+    _ComplexLike,
+    _NumberLike,
+    generic,
+    bool_,
+    timedelta64,
+    number,
+    integer,
+    unsignedinteger,
+    signedinteger,
+    int32,
+    int64,
+    floating,
+    float32,
+    float64,
+    complexfloating,
+    complex128,
+)
+
+if sys.version_info >= (3, 8):
+    from typing import Protocol
+    HAVE_PROTOCOL = True
+else:
+    try:
+        from typing_extensions import Protocol
+    except ImportError:
+        HAVE_PROTOCOL = False
+    else:
+        HAVE_PROTOCOL = True
+
+if HAVE_PROTOCOL:
+    _NumberType = TypeVar("_NumberType", bound=number)
+    _NumberType_co = TypeVar("_NumberType_co", covariant=True, bound=number)
+    _GenericType_co = TypeVar("_GenericType_co", covariant=True, bound=generic)
+
+    class _BoolArithmetic(Protocol[_GenericType_co]):
+        @overload
+        def __call__(self, __other: _BoolLike) -> _GenericType_co: ...
+        @overload  # platform dependent
+        def __call__(self, __other: int) -> Union[int32, int64]: ...
+        @overload
+        def __call__(self, __other: float) -> float64: ...
+        @overload
+        def __call__(self, __other: complex) -> complex128: ...
+        @overload
+        def __call__(self, __other: _NumberType) -> _NumberType: ...
+
+    class _BoolSub(Protocol):
+        # Note that `__other: bool_` is absent here
+        @overload  # platform dependent
+        def __call__(self, __other: int) -> Union[int32, int64]: ...
+        @overload
+        def __call__(self, __other: float) -> float64: ...
+        @overload
+        def __call__(self, __other: complex) -> complex128: ...
+        @overload
+        def __call__(self, __other: _NumberType) -> _NumberType: ...
+
+    class _BoolTrueDiv(Protocol):
+        @overload
+        def __call__(self, __other: _BoolLike) -> float64: ...
+        @overload  # platform dependent
+        def __call__(self, __other: int) -> Union[float32, float64]: ...
```

Comment:
This isn't true, the result is always float64 as `np.result_type(np.int32, np.float32)` is `float64`

Reply:
I think (but haven't checked) the truth is more like
```suggestion
        @overload
        def __call__(self, __other: Union[int8]) -> float16: ...
        @overload
        def __call__(self, __other: Union[int16]) -> float32: ...
        @overload
        def __call__(self, __other: Union[int, int32, int64]) -> float64: ...
```

Reply:
I just did some double a quick test locally and it seems the return type `np.bool_` by `np.integer` division is always `float64`:
```python
In [1]: import numpy as np

In [2]: int_types = [
   ...:     int,
   ...:     np.int8,
   ...:     np.int16,
   ...:     np.int32,
   ...:     np.int64,
   ...:     np.longlong,
   ...:     np.uint8,
   ...:     np.uint16,
   ...:     np.uint32,
   ...:     np.uint64,
   ...:     np.ulonglong
   ...: ]

In [3]: true = np.bool_(True)

In [4]: {type(true / typ(1)) for typ in int_types}
Out[4]: {numpy.float64}

In [5]: {type(typ(1) / true) for typ in int_types}                                                                                                                                              
Out[5]: {numpy.float64}
```

Reply:
Fixed in https://github.com/numpy/numpy/pull/17273/commits/06c745f95439b046714aab0c5a3ee44b56b97bcc.